NeoMathEngine/src/CPU/CpuMathEngineSmallMatricesMultiplyDesc.cpp

```diff
@@ -0,0 +1,141 @@
+	bool resultAdd, bool trans2, bool trans1, int secondRowSize ) const
+{
+#ifdef NEOML_USE_MKL
+	if( firstHeight <= SMMD_MinSize && ( trans2 || ( firstWidth <= SMMD_MinSize && secondWidth <= SMMD_MinSize) ) && resultWidth <= SMMD_MinSize )
+	{
+		return new CCpuSmallMatricesMultiplyDesc( firstHeight, firstWidth, secondWidth, resultWidth, resultAdd, trans2, trans1, secondRowSize );
+	}
+#else  // !NEOML_USE_MKL
+	( void )firstHeight;
+	( void )firstWidth;
+	( void )secondWidth;
+	( void )resultWidth;
+	( void )resultAdd;
+	( void )trans2;
+	( void )trans1;
+	( void )secondRowSize;
+#endif // !NEOML_USE_MKL
+	return nullptr;
+}
+
+inline void CCpuMathEngine::SmallMatricesMultiply( const CSmallMatricesMultiplyDesc& desc,
+	const CConstFloatHandle& firstHandle, const CConstFloatHandle& secondHandle, const CFloatHandle& resultHandle ) const
+{
+	smallMatricesMultiply( desc, GetRaw( firstHandle ), GetRaw( secondHandle ), GetRaw( resultHandle ) );
+}
+
+inline void CCpuMathEngine::smallMatricesMultiply( const CSmallMatricesMultiplyDesc& desc, const float* first, const float* second, float* result ) const
+{
+	const CCpuSmallMatricesMultiplyDesc& mulDesk = static_cast<const CCpuSmallMatricesMultiplyDesc&>( desc );
+	mulDesk.Multiply( first, second, result );
+}
+
+//-------------------------------------------------------------------------------------------------------------------------
+
+CCpuSmallMatricesMultiplyDesc::CCpuSmallMatricesMultiplyDesc( int firstHeight, int firstWidth, int secondWidth, int resultWidth,
+	bool resultAdd, bool trans2, bool trans1, int secondRowSize )
+{
+#ifdef NEOML_USE_MKL
+	secondRowSize = ( secondRowSize > 0 ) ? secondRowSize : secondWidth;
+
+	// Create jitter handle and generate GEMM kernel
+	auto status = mkl_jit_create_sgemm(
+		&MklJitter,
+		MKL_ROW_MAJOR,
+		( trans1 ? MKL_TRANS : MKL_NOTRANS ),
+		( trans2 ? MKL_TRANS : MKL_NOTRANS ),
+		/*m*/( trans1 ? firstWidth : firstHeight ),
+		/*n*/( trans2 ? resultWidth : secondWidth ),
+		/*k*/( trans1 ? firstHeight : firstWidth ),
+		1.f,
+		NEOML_MAX( 1, firstWidth ),
+		NEOML_MAX( 1, secondRowSize ),
+		( resultAdd ? 1.f : 0.f ),
+		NEOML_MAX( 1, resultWidth ) );
+	ASSERT_EXPR( status != MKL_JIT_ERROR );
+
+	// Get kernel associated with jitter handle
+	MklKernel = mkl_jit_get_sgemm_ptr( MklJitter );
+	ASSERT_EXPR( MklKernel != nullptr );
+
+#else  // !NEOML_USE_MKL
+	( void )firstHeight;
+	( void )firstWidth;
+	( void )secondWidth;
+	( void )resultWidth;
+	( void )resultAdd;
+	( void )trans2;
+	( void )trans1;
+	( void )secondRowSize;
+#endif // !NEOML_USE_MKL
+}
+
+CCpuSmallMatricesMultiplyDesc::~CCpuSmallMatricesMultiplyDesc()
+{
+	ASSERT_EXPR( MklJitter != nullptr && MklKernel != nullptr );
+#ifdef NEOML_USE_MKL
+	// Destroy the created jitter/GEMM kernel
+	mkl_jit_destroy( MklJitter );
+	// Avoid memory leak
+	mkl_free_buffers();
```

Comment:
It shouldn't be here. `mkl_free_buffers` will be called at the end of the program (or by special `CpuMathEngineCleanUp()`). It's definitely not a good idea to call it each time jit is destroyed.

Reply:
fixed